ledger/complete/mtrie/trieCache_test.go

```diff
@@ -0,0 +1,134 @@
+	const capacity = 10
+
+	tc := NewTrieCache(capacity, nil)
+	require.Equal(t, 0, tc.Count())
+
+	tries := tc.Tries()
+	require.Equal(t, 0, len(tries))
+	require.Equal(t, 0, tc.Count())
+	require.Equal(t, 0, len(tc.lookup))
+
+	// savedTries contains all tries that are pushed to queue
+	var savedTries []*trie.MTrie
+
+	// Push tries to queue to fill out capacity
+	for i := 0; i < capacity; i++ {
+		trie, err := randomMTrie()
+		require.NoError(t, err)
+
+		tc.Push(trie)
+
+		savedTries = append(savedTries, trie)
+
+		tr := tc.Tries()
+		require.Equal(t, savedTries, tr)
+		require.Equal(t, len(savedTries), tc.Count())
+		require.Equal(t, len(savedTries), len(tc.lookup))
+
+		retTrie, found := tc.Get(trie.RootHash())
+		require.Equal(t, retTrie, trie)
+		require.True(t, found)
+
+		// check last added trie functionality
+		retTrie = tc.LastAddedTrie()
+		require.Equal(t, retTrie, trie)
+	}
+
+	// Push more tries to queue to overwrite older elements
+	for i := 0; i < capacity; i++ {
+		trie, err := randomMTrie()
+		require.NoError(t, err)
+
+		tc.Push(trie)
+
+		savedTries = append(savedTries, trie)
+
+		tr := tc.Tries()
+		require.Equal(t, capacity, len(tr))
+
+		// After queue reaches capacity in previous loop,
+		// queue overwrites older elements with new insertions,
+		// and element count is its capacity value.
+		// savedTries contains all elements inserted from previous loop and current loop, so
+		// tr (queue snapshot) matches the last C elements in savedTries (where C is capacity).
+		require.Equal(t, savedTries[len(savedTries)-capacity:], tr)
+		require.Equal(t, capacity, tc.Count())
+		require.Equal(t, capacity, len(tc.lookup))
+
+		// check the trie is lookable
+		retTrie, found := tc.Get(trie.RootHash())
+		require.Equal(t, retTrie, trie)
+		require.True(t, found)
+
+		// check the last evicted value is not kept
+		retTrie, found = tc.Get(savedTries[len(savedTries)-capacity-1].RootHash())
+		require.Nil(t, retTrie)
+		require.False(t, found)
+
+		// check last added trie functionality
+		retTrie = tc.LastAddedTrie()
+		require.Equal(t, retTrie, trie)
+	}
+
+	// test purge functionality
+	tc.Purge()
+	require.Equal(t, 0, tc.Count())
+	require.Equal(t, 0, tc.tail)
+	require.Equal(t, 0, len(tc.lookup))
+}
+
+func TestEvictCallBack(t *testing.T) {
+	const capacity = 2
+
+	trie1, err := randomMTrie()
+	require.NoError(t, err)
+
+	called := false
+	tc := NewTrieCache(capacity, func(tree *trie.MTrie) {
+		called = true
+		require.Equal(t, trie1, tree)
+	})
+	tc.Push(trie1)
+
+	trie2, err := randomMTrie()
+	require.NoError(t, err)
+	tc.Push(trie2)
+
+	trie3, err := randomMTrie()
+	require.NoError(t, err)
+	tc.Push(trie3)
+
+	require.True(t, called)
+}
+
+func randomMTrie() (*trie.MTrie, error) {
+	var randomPath ledger.Path
+	rand.Read(randomPath[:])
+
+	var randomHashValue hash.Hash
+	rand.Read(randomHashValue[:])
+
+	root := node.NewNode(256, nil, nil, randomPath, nil, randomHashValue)
+
+	return trie.NewMTrie(root, 1, 1)
+}
```

Comment:
Just one simple concurrency test, this is mission-critical piece of code

Reply:
added a simple on at [02fc84c](https://github.com/onflow/flow-go/pull/2893/commits/02fc84c2a4e380356b853eb396fe35cea68db981)